packaging/_tokenizer.py

```diff
@@ -0,0 +1,153 @@
+import re
+from typing import Any, Dict, Generator, Optional, Sized
+
+from .specifiers import Specifier
+
+
+class Token:
+    def __init__(self, name: str, text: str, position: int):
+        self.name = name
+        self.text = text
+        self.position = position
+
+    def matches(self, name: str = "", text: str = "") -> bool:
+        if name and self.name != name:
+            return False
+        return True
+
+
+class ParseException(Exception):
+    """Parsing failed"""
+
+    def __init__(self, message: str, position: int) -> None:
+        super().__init__(message)
+        self.position = position
+
+
+DEFAULT_RULES = {
+    None: r"[ \t]+",  # whitespace: not returned as tokens
+    "LPAREN": r"\(",
+    "RPAREN": r"\)",
+    "LBRACKET": r"\[",
+    "RBRACKET": r"\]",
+    "SEMICOLON": r";",
+    "COLON": r",",
+    "QUOTED_STRING": re.compile(
+        r"""
+            ('[^']*')
+            |
+            ("[^"]*")
+        """,
+        re.VERBOSE,
+    ),
+    "OP": r"===|==|~=|!=|<=|>=|<|>",
+    "VERSION": re.compile(Specifier._version_regex_str, re.VERBOSE | re.IGNORECASE),
+    "BOOLOP": r"or|and",
+    "IN": r"in",
+    "NOT": r"not",
+    "VARIABLE": re.compile(
+        r"""
+            python_version
+            |python_full_version
+            |os[._]name
+            |sys[._]platform
+            |platform_(release|system)
+            |platform[._](version|machine|python_implementation)
+            |python_implementation
+            |implementation_(name|version)
+            |extra
+        """,
+        re.VERBOSE,
+    ),
+    "URL_SPEC": "@ *[^ ]+",
+    "IDENTIFIER": r"([a-zA-Z0-9]|-|_|\.)+",
+}
+
+
+class Tokenizer:
+    """Stream of tokens for a LL(1) parser.
+
+    Provides methods to examine the next token to be read, and to read it
+    (advance to the next token).
+    """
+
+    next_token: Any
+
+    def __init__(
+        self, source: Sized, rules: Dict[Optional[str], object] = DEFAULT_RULES
+    ) -> None:
+        self.source = source
+        self.rules = {name: re.compile(pattern) for name, pattern in rules.items()}
+        self.next_token = None
+        self.generator = self._tokenize()
+        self.position = 0
+
+    def peek(self, *match_args: None, **match_kwargs: None) -> Any:
```

Comment:
These none annotations are weird. If it's not supposed to be passed, just don't have varargs/kwargs?

Reply:
This basically applies to the whole file.